src/main/java/io/vertx/ext/mail/SMTPException.java

```diff
@@ -0,0 +1,90 @@
+/*
+ *  Copyright (c) 2011-2021 The original author or authors
+ *
+ *  All rights reserved. This program and the accompanying materials
+ *  are made available under the terms of the Eclipse Public License v1.0
+ *  and Apache License v2.0 which accompanies this distribution.
+ *
+ *       The Eclipse Public License is available at
+ *       http://www.eclipse.org/legal/epl-v10.html
+ *
+ *       The Apache License v2.0 is available at
+ *       http://www.opensource.org/licenses/apache2.0.php
+ *
+ *  You may elect to redistribute this code under either of these licenses.
+ */
+
+package io.vertx.ext.mail;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * This represents an Exception during communication with SMTP server.
+ *
+ * @author <a href="mailto: aoingl@gmail.com">Lin Gao</a>
+ */
+public class SMTPException extends Exception {
+
+  private final int replyCode;
+  private final List<String> replyMessages;
+
+  /**
+   * Constructor of SMTPException.
+   *
+   * @param message the informative message prepending the reply messages
+   * @param replyCode the SMTP reply code
+   * @param replyMessages the SMTP reply messages
+   */
+  public SMTPException(String message, int replyCode, List<String> replyMessages) {
+    super(message + ": " + String.join("\n", Objects.requireNonNull(replyMessages)));
+    this.replyCode = replyCode;
+    this.replyMessages = replyMessages;
+  }
+
+  /**
+   * Gets the SMTP reply status code.
+   *
+   * @return the SMTP reply status code.
+   */
+  public int getReplyCode() {
+    return replyCode;
+  }
+
+  /**
+   * Gets the SMTP reply messages. It maybe multi lines messages splits by <code>\n</code>
+   *
+   * @return the SMTP reply messages
+   */
+  public List<String> getReplyMessages() {
+    return replyMessages;
+  }
+
+  /**
+   * Gets the SMTP reply message. In case of multi lines messages, returns the first line.
+   *
+   * @return the SMTP reply message
+   */
+  public String getReplyMessage() {
+    return replyMessages.get(0);
+  }
+
+  /**
+   * Checks if it is a permanent failure, that the reply status code <code>&gt;=</code> 500
+   *
+   * @return true if it is a permanent failure, false otherwise.
+   */
+  public boolean isStatusFatal() {
+    return replyCode >= 500;
+  }
+
+  /**
+   * Checks if it is a transient failure, that the reply status code <code>&gt;=</code> 400 and <code>&lt;</code> 500
+   *
+   * @return true if it is a permanent failure, false otherwise.
+   */
+  public boolean isStatusTemporary() {
+    return replyCode >= 400 && replyCode < 500;
+  }
```

Comment:
Is this terminology of "permanent"/"fatal" and "transient"/"temporary" mandated by the spec/RFC?

I find it a bit confusing that something like "Domain not found" (450) is considered non-permanent. But I guess the logic behind this is that a domain could just be "offline"due to DNS problems or similar?

Secondarily, I personally would find it clearer if those synonyms could be avoided (e.g. "permanent"/"fatal"), better stick with a single term for each case?

Reply:
Yes, they are defined at https://datatracker.ietf.org/doc/html/rfc5321#section-4.2.1:
> 4yz  Transient Negative Completion reply
> 5yz  Permanent Negative Completion reply

so the reply codes that `>= 400 and < 499` are considered as transient failures, and the reply codes that `>=500` are considered as permanent failures.

> I find it a bit confusing that something like "Domain not found" (450) is considered non-permanent. But I guess the logic behind this is that a domain could just be "offline"due to DNS problems or similar?

I guess so :)

> Secondarily, I personally would find it clearer if those synonyms could be avoided (e.g. "permanent"/"fatal"), better stick with a single term for each case?

Yes, I will update the methods name to `isPermanent()` and `isTransient()`.